pkg/server/api/localauthority/v1/service.go

```diff
@@ -0,0 +1,279 @@
+		return nil, api.MakeErr(log, codes.InvalidArgument, "invalid authority ID", err)
+	}
+	log = log.WithField(telemetry.LocalAuthorityID, authorityID)
+
+	if err := s.ds.TaintX509CA(ctx, s.td.IDString(), publicKey); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to taint X.509 authority", err)
+	}
+
+	state := &localauthorityv1.AuthorityState{
+		AuthorityId: authorityID,
+		Status:      localauthorityv1.AuthorityState_OLD,
+	}
+
+	rpccontext.AuditRPC(ctx)
+	log.Info("X.509 authority tainted successfully")
+
+	return &localauthorityv1.TaintX509AuthorityResponse{
+		TaintedAuthority: state,
+	}, nil
+}
+
+func (s *Service) RevokeX509Authority(ctx context.Context, req *localauthorityv1.RevokeX509AuthorityRequest) (*localauthorityv1.RevokeX509AuthorityResponse, error) {
+	rpccontext.AddRPCAuditFields(ctx, buildAuditLogFields(req.AuthorityId))
+	log := rpccontext.Logger(ctx)
+
+	authorityID, publicKey, err := s.getX509PublicKey(ctx, req.AuthorityId)
+	if err != nil {
+		if req.AuthorityId != "" {
+			log = log.WithField(telemetry.LocalAuthorityID, req.AuthorityId)
+		}
+		return nil, api.MakeErr(log, codes.InvalidArgument, "invalid authority ID", err)
+	}
+
+	log = log.WithField(telemetry.LocalAuthorityID, authorityID)
+	if err := s.ds.RevokeX509CA(ctx, s.td.IDString(), publicKey); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to revoke X.509 authority", err)
+	}
+
+	state := &localauthorityv1.AuthorityState{
+		AuthorityId: authorityID,
+		Status:      localauthorityv1.AuthorityState_OLD,
+	}
+
+	rpccontext.AuditRPC(ctx)
+	log.Info("X.509 authority revoked successfully")
+
+	return &localauthorityv1.RevokeX509AuthorityResponse{
+		RevokedAuthority: state,
+	}, nil
+}
+
+// getX509PublicKey validates provided authority ID, and return OLD associated publick key, in case of authority ID is not provided, use the current OLD authority
+func (s *Service) getX509PublicKey(ctx context.Context, authorityID string) (string, crypto.PublicKey, error) {
+	if authorityID == "" {
+		// No key provided, taint OLD key
+		nextSlot := s.ca.GetNextX509CASlot()
+		if !nextSlot.IsEmpty() {
+			return "", nil, errors.New("unable to use a prepared key")
+		}
+
+		return nextSlot.AuthorityID(), nextSlot.PublicKey(), nil
+	}
+
+	nextSlot := s.ca.GetNextX509CASlot()
+	if nextSlot.AuthorityID() == authorityID {
+		return nextSlot.AuthorityID(), nextSlot.PublicKey(), nil
+	}
+
+	currentSlot := s.ca.GetCurrentX509CASlot()
+	if currentSlot.AuthorityID() == authorityID {
+		return "", nil, errors.New("unable to use current authority")
+	}
+
+	bundle, err := s.ds.FetchBundle(ctx, s.td.IDString())
+	if err != nil {
+		return "", nil, err
+	}
+
+	for _, ca := range bundle.RootCas {
+		cert, err := x509.ParseCertificate(ca.DerBytes)
+		if err != nil {
+			return "", nil, err
+		}
+
+		subjectKeyID := x509util.SubjectKeyIDToString(cert.SubjectKeyId)
+		if authorityID == subjectKeyID {
+			return subjectKeyID, cert.PublicKey, nil
+		}
+	}
```

Comment:
Should we be searching the bundle for these keys? I think one side effect of this is I can taint a key in the bundle that isn't managed locally ... Since it's the Local Authority API, kind of feels like you should be interacting with the server instance which owns the key? Or are there use cases that require this kind of thing?

Reply:
searching be publickey was my initial idea too, as you said we may be able to taint a no local authority, I dont have a use case about that "yet",
So are you ok with getting authorityID and validate if it is used in the Old authority and reject if that is not the case? (we can have the Old one once we add the third Slot)

Reply:
Yea that sounds good .. probably better to start with local only, and extend it if/when we have a good use case for that